state/relationunit.go

```diff
@@ -475,57 +476,80 @@ func (ru *RelationUnit) ReadSettings(uname string) (m map[string]interface{}, er
+// relation endpoint is bound to a space.
+func NetworksForRelation(
+	binding string, unit *Unit, rel *Relation, defaultEgress []string,
+) (boundSpace string, ingress []string, egress []string, _ error) {
+	st := unit.st
+
+	relEgress := NewRelationEgressNetworks(st)
+	egressSubnets, err := relEgress.Networks(rel.Tag().Id())
+	if err != nil && !errors.IsNotFound(err) {
+		return "", nil, nil, errors.Trace(err)
+	} else if err == nil {
+		egress = egressSubnets.CIDRS()
+	} else {
+		egress = defaultEgress
+	}
+
+	boundSpace, err = unit.GetSpaceForBinding(binding)
+	if err != nil && !errors.IsNotValid(err) {
+		return "", nil, nil, errors.Trace(err)
+	}
+	// If the endpoint for this relation is not bound to a space, or
+	// is bound to the default space, we need to look up the ingress
+	// address info which is aware of cross model relations.
+	if boundSpace == environs.DefaultSpaceName || err != nil {
+		crossmodel, err := rel.IsCrossModel()
+		if err != nil {
+			return "", nil, nil, errors.Trace(err)
+		}
+		if crossmodel {
+			address, err := unit.PublicAddress()
+			if err != nil {
+				// TODO(wallyworld) - it's ok to return a private address sometimes
+				// TODO return an error when it's not possible to use the private address
+				logger.Warningf(
+					"no public address for unit %q in cross model relation %q, using private address",
+					unit.Name(), rel)
+				address, err = unit.PrivateAddress()
+				if err != nil {
+					return "", nil, nil, errors.Trace(err)
+				}
+			}
+			ingress = []string{address.Value}
+		}
 	}
-	if crossmodel, err := ru.relation.IsCrossModel(); err != nil {
-		return network.Address{}, errors.Trace(err)
-	} else if !crossmodel {
-		space, err := unit.GetSpaceForBinding(ru.endpoint.Name)
+	if len(ingress) == 0 {
+		// We don't yet have an ingress address, so pick one from the space to
+		// which the endpoint is bound.
+		machineID, err := unit.AssignedMachineId()
 		if err != nil {
-			return network.Address{}, errors.Trace(err)
+			return "", nil, nil, errors.Trace(err)
 		}
-		m, err := unit.machine()
+
+		machine, err := st.Machine(machineID)
 		if err != nil {
-			return network.Address{}, errors.Trace(err)
+			return "", nil, nil, errors.Trace(err)
 		}
-		networkInfos := m.GetNetworkInfoForSpaces(set.NewStrings(space))
-		info, ok := networkInfos[space]
-		if !ok || len(info.NetworkInfos) == 0 || len(info.NetworkInfos[0].Addresses) == 0 {
-			privateAddress, err := unit.PrivateAddress()
-			if err == nil {
-				logger.Warningf("Can't find an address for default binding space %q, falling back to units' private address %q", space, privateAddress.String())
+
+		networkInfos := machine.GetNetworkInfoForSpaces(set.NewStrings(boundSpace))
+		// The binding address information based on link layer devices.
+		for _, nwInfo := range networkInfos[boundSpace].NetworkInfos {
+			for _, addr := range nwInfo.Addresses {
+				ingress = append(ingress, addr.Address)
```

Comment:
you got rid of falling back to PrivateAddress in the case where networkInfos is length 0, is that intentional?

Reply:
Yeah. If it's a cross model relation on an endpoint bound to the default space, we still fall back to private address for the ingress address.
Otherwise, we rely on  the behaviour in GetNetworkInfoForSpaces() to "Do The Right Thing" with regard to populating the bind addresses. And we grab the first of those to use as the ingress address.